### ISSUE TITLE
FEATURE REQUEST: Support GCE Internal Load Balancer

### ISSUE THREAD

**Issue by tafypz** [kubernetes/kubernetes#33483]
I would be great to be able to use GCE new internal load balancer feature.
Configuration could be done via the same type of config as AWS internal load balancer.
This would avoid having to create bastion routes to access non public services living in GKE.

**Comment by pdecat** [comment]
FWIW, GCE Internal Load Balancers made it to the compute API v1 since november 21st: https://github.com/google/google-api-go-client/blob/master/compute/v1/compute-gen.go#L1803

**Comment by thockin** [comment]
Can you tell me how you're hoping to use ILB?  There are some issues with
it that I have not yet worked through wrt Kubernetes, and we're looking for
more user input on what result you hope to achieve.

**Comment by tafypz** [comment]
@thockin The usage is fairly simple, we have services that need to be exposed outside a kubernetes cluster (we use container engine). Using the typical load balancer exposes a public IP which in our case is not necessary. We used to use the kubernetes on AWS and used the AWS internal load balancers in our service config. So all in all simple usage, access kubernetes services from outside a container engine cluster and expose the services to compute engine instances without exposing them to the public. Right now we create bastion routes to access k8s services this but it is not ideal.

**Comment by thockin** [comment]
Thanks.  That fits one of the expected usage modes, and actually one of the easier to implement ones.  Truth is that nobody is working on it just now, but we're planning for it.

**Comment by pdecat** [comment]
Same usecase here.

**Comment by rvrignaud** [comment]
Very same use case here. For now we are adding static routes for servicesIpv4Cidr to cluster nodes this is really ugly.

**Comment by kuroneko25** [comment]
Exact same use case. We are on GKE and currently uses bastion routes. AFAIK that is the best alternative at this point.

**Comment by ipadavic** [comment]
Same use case. Need to access GKE services from GCE instances.

**Comment by replicant0wnz** [comment]
+1

**Comment by evaldasou** [comment]
no updates on this one?

**Comment by thockin** [comment]
We're working on a couple of things that need to be done before this can
really be done, and working through the limitations.

**Comment by vdm** [comment]
Out of the box:
1. GKE pods can ping GCE by IP and by name, but not vice versa.
2. GCE VMs can ping GKE pods by IP (obtained from `ip addr` inside the pod), but not by name. This uses a GCE route generated by GKE to a "VMs" /24 subnet.

A GCE route to a subnet including generated K8s Service IPs can be added manually, with the same next hop as the generated route. This allows GCE VMs to access GKE services IPs (but not by name) without using a GCE internal load balancer. Now it is possible to use statically configured hostnames on GCE pointing to K8s Service IPs.

Instead of exposing GCE internal load balancers, GKE could create a GCE route for Services and publish Pod/Service names to the GCE metadata/DNS server.

**Comment by thockin** [comment]
The downside of this technique (we call it service bastion routes) is that
if the node to which you are routing goes down, the service gateway is
down.  It needs a rectifier to check and choose a new node or nodes when it
fails, if you want to offer any sort of availability SLA.

**Comment by kuroneko25** [comment]
How much integration are we currently planning to have between ILB and Kubernetes/GKE? Will the ILB be able to route to a specific service VIP or pod? Or is it only able to forward request to a host port and we need to have an intermediate proxy layer running in the cluster ? i.e. nginx, haproxy, envoy etc.

**Comment by GrantGochnauer** [comment]
+1

**Comment by edenxia** [comment]
we have the same requirement. couple of scenarios. though, we can setup one cluster across zones , we still need db replication and internal service call across regions

**Comment by denisa** [comment]
+1

**Comment by jlewi** [comment]
+1

**Comment by itamaro** [comment]
Not directly related to ILB, but the new [Identity-Aware Proxy](https://cloud.google.com/iap/) could solve the same use-case, so as far as I'm concerned - whichever of these integrates with GKE first wins :-)

[API for IAP](https://cloud.google.com/iap/docs/reference/compute-engine-apis)

**Comment by peay** [comment]
I have the same use case too. Internal load balancers would be great, although automatic generation of the service bastion route and exporting service names to the metadata/DNS server as mentioned by @vdm would be even better.

This would greatly enhance our ability to interface services in GKE with other GCP products easily, especially in a CI/CD environment where there can be many namespaces and services.

**Comment by bviolier** [comment]
+1 We would primarily use it as a solution to switch services from on-premise to GCP (through VPN between the two).

**Comment by bviolier** [comment]
Would another probable solution be to VPN through Google VPN into the cluster (with openVPN service for instance) and in that way make the ClusterIP available?

Do note that you then also need to have a non-changing ClusterIP or also expose the DNS so you can use the service name.

**Comment by dgpc** [comment]
FYI for the folks talking about using VPN and the Internal Load Balancer:
The documentation on internal load balancers specifically states that it cannot receive traffic through a VPN tunnel, it can only receive traffic from GCE nodes in the same region.

> You cannot send traffic through a VPN tunnel to your load balancer IP.

https://cloud.google.com/compute/docs/load-balancing/internal/

We ran into this problem, and worked around it by setting up NodePort type Services, then configuring BIND on our non-GCP machines to forward requests to Kube DNS. That way applications outside GCP could resolve the Node IPs, and connect to them via the VPN tunnel.

**Comment by bviolier** [comment]
@dgpc Thanks for the info! The VPN for us would be a temporary solution, and if the GCP VPN's don't work, we could also go for some GCE instances that take on this role :-) 

Happy to see that the issue is assigned and being worked on.

**Comment by thockin** [comment]
For this who are waiting for ILB - how often do you need BOTH an internal
ILB and an external LB on the same Service?

**Comment by tafypz** [comment]
@thockin Our usage is such that we would not require both external and internal on the same service.
If having both slows down the development, i would vote for having only one type LB per service first and both as an enhancement later.

**Comment by pdecat** [comment]
Hi @thockin, in our current setup, we do not need both Internal and External LB on the same services.

We are using ILB to expose non public facing services from one K8S cluster to another, both in distinct subnets.

**Comment by itamaro** [comment]
> For this who are waiting for ILB - how often do you need BOTH an internal
> ILB and an external LB on the same Service?

we have **no** use-case for both ILB and external LB on the same service.

we need external LB's for externally accessible services, and ILB's for simple & direct access to non-public-facing services (AKA "internal services", "dev/test stack", etc.) from outside the GKE cluster, from personal dev machines / office network.

**Comment by rvrignaud** [comment]
We have same use case as @itamaro so we don't need both internal and external LB for one particular service

**Comment by rnavarro** [comment]
@thockin Same as @itamaro 

We don't have any use cases where we deploy both a public and internal LB for a single service.

**Comment by kuroneko25** [comment]
For us a service is either internal or external but not both at the same time. So we don't have a case where a service needs to use both ILB and GLBC.

**Comment by davidquarles** [comment]
+1, and same as the previous commenters -- our only use case is internal.

**Comment by jeffyecn** [comment]
+1, only use case is internal.

**Comment by EmiPhil** [comment]
+1 only use case is internal

**Comment by jr-qordoba** [comment]
+1 for internal only

**Comment by jlaham** [comment]
+1 for separate intLB and extLB services.

If anyone does need both intLB and extLB, wouldn't it be as straight forward as defining two separate services (one intLB, and one extLB), both pointing to the same deployment?

**Comment by ualtinok** [comment]
+ for internal only

**Comment by slavakl** [comment]
+ 1 for separate intLB and extLB services, totally agree on with @jlaham.

@thockin Do you have any update or ETA on this?

**Comment by nicksardo** [comment]
ILB setup via service controller is under development and should be shipped in 1.7

**Comment by slavakl** [comment]
thank you, @nicksardo.

**Comment by roobert** [comment]
@nicksardo: any word on whether ILBs will be accessible via GCP VPNs?

**Comment by replicant0wnz** [comment]
@roobert I like this question. Currently internal LB's can't be accessed via VPN's or other projects :-( We currently have to roll our own HAProxy deployments for this very reason ..

**Comment by pires** [comment]
@nicksardo is there a PR or different issues one can subscribe to?

Also, can you fix the labels for this issue?

**Comment by myaghini** [comment]
Same case, same use. Need internal lb for container cluster services to access them from VM and from VPN.

**Comment by astraverkhau** [comment]
Same requirement here.

**Comment by nicksardo** [comment]
How many internal load balancers do all of you estimate you'll run per K8s cluster (or region)? Keep in mind that this is only for exposing K8s services externally of the K8s cluster but within the same GCP subnetwork. 

Appreciate your responses, thanks.

**Comment by astraverkhau** [comment]
I would  roughly estimate 2/3 of our 360 (and counting) micro services will need to be available strictly on GCP network and not exposed outside, so 240+ per cluster.

**Comment by itamaro** [comment]
> How many internal load balancers do all of you estimate you'll run per K8s cluster/region? Keep in mind that this is only for exposing K8s services externally of the K8s cluster but within the same GCP subnetwork.

I'd estimate it at around 50 for the near future for our environment.

**Comment by roobert** [comment]
@nicksardo - will the internal load balancers be routable via GCP VPNs?

**Comment by pires** [comment]
Why was this closed?

**Comment by evanj** [comment]
@nicksardo I think for us (Bluecore Inc), we would only use 2 internal load balancers immediately, and maybe a total of about 5 when I think about projects that are currently in flight over say the next year.

**Comment by nicksardo** [comment]
Beta support for ILBs have been added in #46663 and will be shipped with 1.7. Documentation on how to enable it will come at release. 

Thanks @evanj @itamaro @astraverkhau for your data points.

**Comment by pmazzuca** [comment]
Just as a clarification, when will Google Container Engine support 1.7, and the internal load balancer?  And will an organization using the Google VPN to a VPC be able to access the IP of that internal load balancer?

**Comment by roobert** [comment]
@nicksardo: I can't find anything in the 1.7 changelog about this feature, has it been postponed?

It would also be great to get clarification on whether the internal load balancers will be accessible to networks attached via VPNs.

Thanks,

**Comment by nicksardo** [comment]
It's in the changelog: "[beta] Support creation of GCP Internal Load Balancers from Service objects " https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG.md#cloud-provider-enhancement

Folks, I can only comment on integrating ILBs into Kubernetes. For everything else, you have to go through the correct channels: https://cloud.google.com/support/docs/issue-trackers.

**Comment by roobert** [comment]
After some further searching it looks like the answer to whether internal load balancers are routable via GCP VPNs can be found under the _Restrictions_[1] heading: "You cannot send traffic through a VPN tunnel to your load balancer IP."

Hope this helps the others.

[1] https://cloud.google.com/compute/docs/load-balancing/internal/

**Comment by kuroneko25** [comment]
Are there any docs published?

**Comment by EmiPhil** [comment]
For anyone looking for docs on this, I've found the Google Kubernetes ILB docs to be sufficient: [https://cloud.google.com/container-engine/docs/internal-load-balancing](https://cloud.google.com/container-engine/docs/internal-load-balancing)

**Comment by nicksardo** [comment]
There's a port collision between the kube-proxy health check server and the node problem detector. As a result, health checks for ILB fail unless the service has externalTrafficPolicy: Local. https://github.com/kubernetes/kubernetes/issues/49263
Sounds like this will be fixed in 1.7.2.

**Comment by cb372** [comment]
I'm trying to set up a service with an ILB in a custom subnet. Is this possible?

Here's my service.yaml:

```
apiVersion: v1
kind: Service
metadata:
  name: my-app
  annotations:
    cloud.google.com/load-balancer-type: "internal"
spec:
  type: LoadBalancer
  ports:
    - port: 9000
  selector:
    app: my-app
```

This fails with the following error:

```
Error creating load balancer (will retry): Failed to create load balancer for service default/dummy-ovo-site: googleapi: Error 400: Invalid value for field 'resource.subnetwork': ''. Network interface must specify a subnet if the network resource is in custom subnet mode., invalid
```

Is there an annotation or something I can add to specify the subnet?

**Comment by nicksardo** [comment]
https://github.com/kubernetes/kubernetes/issues/50034 GKE on K8s 1.7.4 will have this fixed.

**Comment by beast** [comment]
Hi, all. Not sure if this is relevant but I am trying to setup ILB in Gcloud container engine with the quick start hello-web example. Docs I followed: [ILB](https://cloud.google.com/container-engine/docs/internal-load-balancing), [Quick Start](https://cloud.google.com/container-engine/docs/quickstart) 

When I follow the quick example to have an external load balancer created, everything is working fine. However when I follow the ILB article and try to access the ILB from a compute engine SSHusing curl in the same subnet. It's not working. I asked a question on Stackoverflow with detailed description [here](https://stackoverflow.com/questions/45912802/how-do-i-make-app-engine-flexible-to-connect-to-container-engine-via-internal-ip).

Has anyone able to create ILB successfully with GKE and able to access from a compute engine or appengine flexible using internal IP. Any response is greatly appreciated. Thanks.

**Comment by BezVezeE** [comment]
Hi there, i tried to setup ILB on gke to expose internal address on the network so i cann access the app from gce to gke, but the thing is i cant ping the expoesed address from gce node,

any ideas??

**Comment by bviolier** [comment]
@BezVezeE Are you trying to access the app from a GCE instance in the same project AND the same region? Because ILB's only work in the same project & region.

**Comment by BezVezeE** [comment]
@bviolier the gce node is on us-central1-f and the kubernetes cluster is on  us-central1-a ... 
they are in the same region and also they are on the same project.

here is the yaml file of the loadbalancer

`apiVersion: v1
kind: Service
metadata:
  annotations:
    cloud.google.com/load-balancer-type: Internal
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"Service","metadata":{"annotations":{"cloud.google.com/load-balancer-type":"Internal"},"labels":{"name":"some-redis"},"name":"redis","namespace":"default"},"spec":{"ports":[{"port":9000,"protocol":"TCP"}],"selector":{"name":"some-redis"},"type":"LoadBalancer"}}
  creationTimestamp: 2017-10-30T12:36:02Z
  labels:
    name: some-redis
  name: redis
  namespace: default
  resourceVersion: "1033325"
  selfLink: /api/v1/namespaces/default/services/redis
  uid: e3f91581-bd6e-11e7-b112-42010a800135
spec:
  clusterIP: 10.19.248.53
  externalTrafficPolicy: Cluster
  ports:
  - nodePort: 32188
    port: 9000
    protocol: TCP
    targetPort: 9000
  selector:
    name: some-redis
  sessionAffinity: None
  type: LoadBalancer
status:
  loadBalancer:
    ingress:
    - ip: 10.128.0.4`

is there any firewall configuration that im maybe missing out?

**Comment by relvan1** [comment]
Hi.. Is it possible to access the services which is in other cluster(central region) from my cluster(west region) using the ILB? 

Both cluster falls under the same project but different region.
